File: src/test/java/de/fraunhofer/aisec/crymlin/FindingDescriptionTest.java

package de.fraunhofer.aisec.crymlin;

import de.fraunhofer.aisec.analysis.structures.Finding;
import de.fraunhofer.aisec.analysis.structures.FindingDescription;
import de.fraunhofer.aisec.cpg.sarif.Region;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Tests the SARIF-like format for Finding outputs.
 */
public class FindingDescriptionTest {

	@Test
	public void testFindingDescription() throws URISyntaxException {
		// Create some Finding object
		String logMsg = "Variable cm not initialized";
		URI artifactUri = URI.create("file:///tmp/test.cpp");
		String onFailId = "WrongUseOfBotan_CipherMode";
		List<Region> regions = List.of(new Region(1, 2, 10, 12));
		boolean isProblem = true;
		Finding f = new Finding(logMsg, artifactUri, onFailId, regions, isProblem);

		// Initialize database with explanations
		FindingDescription fd = FindingDescription.getInstance();
		fd.init(new File(this.getClass().getClassLoader().getResource("unittests/findingDescription.json").toURI()));

		String fullDescription = fd.getDescriptionFull(onFailId);
		String shortDescription = fd.getDescriptionShort(onFailId);
		String helpUri = fd.getHelpUri(onFailId);
		List<String> fixes = fd.getFixes(onFailId);

		assertEquals("Full description", fullDescription);
		assertEquals("The order of called Botan methods is wrong.", shortDescription);
		assertEquals("https://www.codyze.io/explanations/10", helpUri);
		assertEquals(1, fixes.size());
		assertEquals("Just fix it!", fixes.get(0));

	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/FindingDescriptionTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.analysis.Finding
import de.fraunhofer.aisec.codyze.analysis.FindingDescription.Companion.instance
import de.fraunhofer.aisec.cpg.sarif.Region
import de.fraunhofer.aisec.mark.markDsl.Action
import java.io.File
import java.net.URI
import java.net.URISyntaxException
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import org.junit.jupiter.api.Test

/** Tests the SARIF-like format for Finding outputs. */
internal class FindingDescriptionTest {
    @Test
    @Throws(URISyntaxException::class)
    fun testFindingDescription() {
        // Create some Finding object
        val logMsg = "Variable cm not initialized"
        val artifactUri = URI.create("file:///tmp/test.cpp")
        val id = "WrongUseOfBotan_CipherMode"
        val regions = listOf(Region(1, 2, 10, 12))
        val isProblem = true
        Finding(id, Action.FAIL, logMsg, artifactUri, regions, isProblem)

        // Initialize database with explanations
        val fd = instance
        val url = this.javaClass.classLoader.getResource("unittests/findingDescription.json")

        assertNotNull(url)

        fd.init(File(url.toURI()))

        val item = fd.get(id)
        val fullDescription = fd.getDescriptionFull(id)
        val shortDescription = fd.getDescriptionShort(id)
        val passMessage = fd.getDescriptionPass(id)
        val helpUri = fd.getHelpUri(id)
        val fixes = fd.getFixes(id)

        assertNotNull(item)
        assertEquals("Full description", fullDescription)
        assertEquals("The order of called Botan methods is wrong.", shortDescription)
        assertEquals("Order of Botan methods validated.", passMessage)
        assertEquals("https://www.codyze.io/explanations/10", helpUri)
        assertEquals(1, fixes?.size)
        assertEquals("Just fix it!", fixes?.get(0) ?: "")
    }
}
